solana/program_v2/programs/gateway_v2/src/instructions/network/gatekeeper_withdraw.rs

```diff
@@ -1,31 +1,62 @@
 use crate::constants::GATEKEEPER_SEED;
 use crate::state::Gatekeeper;
 use anchor_lang::prelude::*;
+use anchor_spl::token::{Mint, Token, TokenAccount, Transfer};
 
 // Will withdraw funds from the gatekeeper
-pub fn gatekeeper_withdraw(ctx: Context<GatekeeperWithdrawAccount>) -> Result<()> {
+pub fn gatekeeper_withdraw(ctx: Context<GatekeeperWithdrawAccount>, mut amount: u64) -> Result<()> {
+    // Amount 0 means withdraw all
+    if amount == 0 {
+        amount = ctx.accounts.gatekeeper_token_account.amount;
+    }
     let gatekeeper = &mut ctx.accounts.gatekeeper;
-    let authority = &mut ctx.accounts.authority;
-    let receiver = &mut ctx.accounts.receiver;
+    let spl_token_program = &mut ctx.accounts.spl_token_program;
+    let receiver_token_account = &mut ctx.accounts.receiver_token_account;
+    let gatekeeper_token_account = &mut ctx.accounts.gatekeeper_token_account;
 
-    gatekeeper.gatekeeper_withdraw(receiver, authority)?;
+    let gatekeeper_authority_key = gatekeeper.authority.key();
+    let gatekeeper_bump = gatekeeper.gatekeeper_bump.to_le_bytes();
+
+    let authority_seed = &[
+        GATEKEEPER_SEED.as_ref(),
+        gatekeeper_authority_key.as_ref(),
+        gatekeeper.gatekeeper_network.as_ref(),
+        gatekeeper_bump.as_ref(),
+    ][..];
+
+    let transfer_instruction = Transfer {
+        from: gatekeeper_token_account.to_account_info(),
+        to: receiver_token_account.to_account_info(),
+        authority: gatekeeper.to_account_info(),
+    };
+
+    let signer = &[authority_seed][..];
+    let cpi_ctx = CpiContext::new_with_signer(
+        spl_token_program.to_account_info(),
+        transfer_instruction,
+        signer,
+    );
+
+    anchor_spl::token::transfer(cpi_ctx, amount)?;
 
     Ok(())
 }
 
-#[derive(Accounts, Debug)]
-#[instruction()]
+#[derive(Accounts)]
+#[instruction(amount: u64)]
 pub struct GatekeeperWithdrawAccount<'info> {
     #[account(
-        mut,
-        seeds = [GATEKEEPER_SEED, authority.key().as_ref()],
-        bump = gatekeeper.gatekeeper_bump,
+        seeds = [GATEKEEPER_SEED, authority.key().as_ref(), gatekeeper.gatekeeper_network.key().as_ref()],
+        bump = gatekeeper.gatekeeper_bump
     )]
     pub gatekeeper: Account<'info, Gatekeeper>,
-    #[account(mut)]
+    #[account()]
     pub authority: Signer<'info>,
-    /// CHECK: Receiver destination account does not need to satisfy the any constraints.
-    #[account(mut)]
-    pub receiver: UncheckedAccount<'info>,
     pub system_program: Program<'info, System>,
+    pub spl_token_program: Program<'info, Token>,
+    pub mint_account: Account<'info, Mint>,
```

Comment:
We should add a comment here that verification is done in CODE. @JulianSpring or do you think there's an elegant way to do it here?

Reply:
I'll add a comment and think of a better solution! Thanks